src/client.ts

```diff
@@ -19,19 +21,28 @@ import {
   applyClientProtocol,
   PartialMsg
 } from './messages/LabeledMsg';
+import { KeyData } from './messages/Lifecycle';
 import {
   EnvData,
   EnvDataHandler,
   LabeledEnvInit,
   Lifecycle
 } from './messages/Lifecycle';
-import { KeyData } from './messages/Lifecycle';
 import { NavRequest } from './messages/NavRequest';
 import { Notification } from './messages/Notification';
 import { Publication } from './messages/Publication';
 
+/**
+ * Registers custom elements used by the client application
+ */
+function registerCustomElementsClient(): void {
+  customElements.define('ifc-client-link', IfcClientLinkElement);
+  customElements.define('ifc-host-link', IfcHostLinkElement);
+}
```

Comment:
Instead of making `registerCustomElementsClient` a standalone function, we can make a `registerCustomElements` method on the `Client` class.  I think this is key to solving most of the issues you're dealing with.  It moots the export issues because there's no new export, and now you have access to a client instance when registering the custom elements.

I'd then update the IfcClientLinkElement to take a client instance as a constructor argument and in the 
`registerCustomElements` method, extend that class, passing the client instance via `super`. Something like this, but maybe with better class names: 

```typescript
registerCustomElements(): void {
  const clientInstance = this;
  class IfcClintLinkElementComplete extends IfcClientLinkElement {
     constructor() {
         super(clientInstance);
     }
  }

 customElements.define('ifc-client-link', IfcClientLinkElementComplete);
}
```